### PR TITLE
Evict containers before removing via V2 API

### LINKED ISSUE #7535: APIv2: DELETE libpod/containers/foo?force=true: unlinkat: directory not empty
Probable flake; filing so we can track it. In API tests:
```
ok 256 [25-containersMore] POST libpod/containers/foo/mount [] : status=200
ok 257 [25-containersMore] Check container mount ('/tmp/test-apiv2.tmp.QENVh1/overlay/7a4337c7d7df8e5d442d365fb12fc4feba80a722bf7775d1416acc001cfe92d2/merged') ~ .*merged
ok 258 [25-containersMore] POST libpod/containers/foo/unmount [] : status=204
not ok 259 [25-containersMore] DELETE libpod/containers/foo?force=true : status
#  expected: 204
#    actual: 500
  expected: 204
not ok 260 [25-containersMore] DELETE libpod/containers/foo?force=true: 204 status returns no output
#  expected: ''
#    actual: {"cause":"1 error occurred:
	* unlinkat /tmp/test-apiv2.tmp.QENVh1/overlay/7a4337c7d7df8e5d442d365fb12fc4feba80a722bf7775d1416acc001cfe92d2: directory not empty
         
","message":"error removing container 41d043ac3e04c3988f10f94be3b457fe2f79f058505ea7c0ad88ecaa0b6123d8 root filesystem: 1 error occurred:
	* unlinkat /tmp/test-apiv2.tmp.QENVh1/overlay/7a4337c7d7df8e5d442d365fb12fc4feba80a722bf7775d1416acc001cfe92d2: directory not empty
         
","response":500}
  expected: ''
```
Source: [apiv2_test](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/5373175691214848/html/apiv2_test.log.html#t--00258) in #7534

Am assuming it's a flake, since I didn't touch any of this code. Will restart failing test.

### PR DESCRIPTION
**Pull request by jwhonce**
Code ported from abi container remove

Fixes #7535

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
@jwhonce I'm not sure the tests are buying into the change.

**Comment by edsantiago** [comment]
Restarted the ubuntu failure, it looks like #7583. The bindings failure, though, looks real.

File: pkg/api/handlers/compat/containers.go [hunk 1]
```diff
@@ -44,11 +45,17 @@ func RemoveContainer(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil && errors.Cause(err) == define.ErrNoSuchCtr {
+		if _, err := runtime.EvictContainer(r.Context(), name, query.Vols); err != nil {
+			if errors.Cause(err) == define.ErrNoSuchCtr {
+				logrus.Debugf("Ignoring error (--allow-missing): %q", err)
+				w.WriteHeader(http.StatusNoContent)
+			} else {
+				utils.InternalServerError(w, err)
+			}
+			return
+		}
```
Comment (edsantiago) [inline pkg/api/handlers/compat/containers.go:57]: Ah, there's a fallthrough here when `err != nil` and the newly-added `Cause` check fails.

File: pkg/api/handlers/compat/containers.go [hunk 2]
```diff
@@ -44,11 +45,24 @@ func RemoveContainer(w http.ResponseWriter, r *http.Request) {
...
 	con, err := runtime.LookupContainer(name)
-	if err != nil {
-		utils.ContainerNotFound(w, name, err)
-		return
-	}
+	if err != nil && errors.Cause(err) == define.ErrNoSuchCtr {
+		// Failed to get container. If force is specified, get the container's ID
+		// and evict it
+		if !query.Force {
+			utils.ContainerNotFound(w, name, err)
```
Comment (TomSweeneyRedHat) [inline pkg/api/handlers/compat/containers.go:52]: idk, but might it be worth a debug statement here as a bread crumb?

Comment (jwhonce) [reply pkg/api/handlers/compat/containers.go:52]: This mirrors the behavior from the ABI code path. I am looking to bring them together not more different.

File: pkg/api/handlers/compat/containers.go [hunk 3]
```diff
@@ -44,11 +45,24 @@ func RemoveContainer(w http.ResponseWriter, r *http.Request) {
...
+			return
+		}
 
+		if _, err := runtime.EvictContainer(r.Context(), name, query.Vols); err != nil {
```
Comment (mheon) [inline pkg/api/handlers/compat/containers.go:56]: I'm still fairly certain the `err == nil` case isn't handled properly - we should be returning immediately with a 204 in that case.

**Comment by mheon** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM
fingers crossed for happy tests.

**Pull request merged**